MSC3784: Using room type of `m.policy` for policy rooms

FSG-Cat opened the pull request:
[Rendered](https://github.com/FSG-Cat/matrix-doc/blob/FSG-Cat-Policy-List-Room-Type/proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md)

This simple MSC aims to help stakeholders know if its likely to be a room containing a policy list or not. 

Signed-off-by: Catalan Lover <catalanlover@protonmail.com>

----

SCT Stuff

[MSC checklist](https://github.com/matrix-org/matrix-spec-proposals/pull/3784#issuecomment-4114179097)

[FCP tickyboxes](https://github.com/matrix-org/matrix-spec-proposals/pull/3784#issuecomment-4114179807)

FSG-Cat committed:
Add files via upload

FSG-Cat committed:
Rename MSCXXXX Using room type of m.policy for policy rooms.md to XXXX-Using-room-type-of-m.policy-for-policy-rooms.md

FSG-Cat committed:
Rename XXXX-Using-room-type-of-m.policy-for-policy-rooms.md to XXXX-using-room-type-of-m.policy-for-policy-rooms.md

FSG-Cat committed:
Update and rename XXXX-using-room-type-of-m.policy-for-policy-rooms.md to 3784-using-room-type-of-m.policy-for-policy-rooms.md

FSG-Cat commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,63 @@
+# MSC3784 Using room type of m.policy for policy rooms
```
Well the way i personally see them used is as dedicated rooms mostly. This proposal aims to help that usecase but is not mutually exclusive with the usecases you mentioned in my head. Yes its probably a bit of a misuse of the type attribute in the sense that policy lists can contain other information unlike the other examples.

Gnuxie replied:
>because there's zero practical reason to do so

@turt2live You're being a bit too certain about this :disappointed:. 

>, it was originally intended that Matrix HQ itself would be the "ban list" plus regular chat, making it easy for others to subscribe to the list and see what was going on. 

Original intent is fair to raise for context, but this isn't how policies are used, and it would be bad to use Matrix HQ this way. Not only is HQ a difficult room to query room state from because of the number of state events it has, but it's just not an appropriate place for them. 

>and see what was going on

Which would also be difficult because policy changes would be mixed with the chat timeline right?

> Policies are just arbitrary state events though and can live anywhere. It's very much a feature that rooms can contain multiple kinds of information when possible, which means a space which is also a policy room (for example).

This isn't how they are used though, even matrix.org has dedicated rooms with chat disabled.  

It makes a lot of sense to show these dedicated rooms differently in clients so that people can browse the rules and admins can edit them.

clokep replied:
Sounds like this makes a good argument and this info just needs to make its way back into the MSC of why the current status quo isn't enough (or just isn't used). It is usually good to reference historical context too.

deepbluev7 commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,63 @@
+
+This proposal is well quite simple. Allow for Policy rooms to be marked as such using the 
+`"type": "m.policy"`. The precedent for this comes from Spaces where they introduced this system in a
+limited capacity. This proposal expands this system to help all stakeholders quickly identify 
+policy rooms in a machine compatible way that is computationally cheap. It has additional benefits like
+allowing clients that are capable of editing policy to display editing tools for policy rooms when they
+detect that a room is a policy room using this mechanic. For machine interaction with policy rooms this
+proposal supplies a very fast way to tell if a room is definitively supposed to be a policy room or if 
+the user might have supplied a legacy room or typed in the wrong room ID / alias depending on how things
+are configured. 
```
To expand on this a bit, I really need something like this. Without this you run into a bootstrapping issue. Is this room supposed to be a policy room before the first policy is sent?

Without a room type you need to inspect the full room state to look for policy rules before you switch to the policy editor design. Slash commands can't warn the user properly that they are sending the rule into the wrong room.

So yes, I very much need this to provide a nice experience in my client. I can somewhat paper over it using account data, but it is not great. Everywhere we are told rooms are cheap. So far no rooms send policy rules into rooms you can send messages in. Most moderation tools even make it impossible to message in policy rooms, since that could possibly break the policy room when you actually need it to fight spam.

Thank you for writing this MSC!

FSG-Cat committed:
Improve Post merge information

Added improved information about what to do after the MSC gets merged.

FSG-Cat commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,65 @@
+# MSC3784 Using room type of m.policy for policy rooms
```
Making this thread to track Implementation status outside of the PR descript. 

Implementation Tracking

Mjolnir as of [da08432](https://github.com/matrix-org/mjolnir/commit/da084328a9dbb12049dce8d8744b82711f8aa881) and that commit is included in [Release 1.6.0](https://github.com/matrix-org/mjolnir/releases/tag/v1.6.0)
Nheko in mtxclient has added initial support to recognise it in [4bd39bce](https://nheko.im/nheko-reborn/mtxclient/-/commit/4bd39bcedd6943ebd87faeac851bbf29b1167c07). Further progres unknown at time of writing.

clokep replied:
Does Nheko use it besides just declaring the constant?

FSG-Cat replied:
Policy Editing that utilises MSC3784 to aid in detecting if a room is a policy list is implemented in RMU. Version that implemented first is unknown but current Version as of writing relies on this as one of its 2 detection methods. It also supports legacy detection.

turt2live reviewed (approved):
My general feeling is this is something we should have in the spec to identify rooms dedicated to containing policy list recommendations. There might still be some rooms which want to mix recommendations with conversation - those rooms wouldn't use this type and might not get the benefits of dedicated UX from a client, but that's okay.

I've given this MSC a sizable editorial review. It's not my intention to completely change how this proposal works, so if that's happened then please correct me 😅. The editorial comments here reflect that ~650 MSCs have been opened in the time since this one was, and that modern MSCs are typically held to a different standard than older ones. I've also tried to make some areas a bit more clear & concise, as would happen in normal review.

My approval denotes that the MSC is ready for proposed-FCP with the edits (or something reasonably close). The listed implementations appear to satisfy the requirements we'd have for this proposal.

FSG-Cat committed:
Apply suggestions made by Travis R.

Co-authored-by: Travis Ralston <travpc@gmail.com>

turt2live commented:
MSCs proposed for Final Comment Period (FCP) should meet the requirements outlined in the [checklist](https://github.com/matrix-org/matrix-spec-proposals/blob/main/MSC_CHECKLIST.md) prior to being accepted into the spec. This checklist is a bit long, but aims to reduce the number of follow-on MSCs after a feature lands. 

SCT members: please check off things you check for, and raise a concern against FCP if the checklist is incomplete. If an item doesn't apply, prefer to check it rather than remove it. Unchecking items is encouraged where applicable.

MSC authors: feel free to ask *in a thread* on your MSC or in the[#matrix-spec:matrix.org](https://matrix.to/#/#matrix-spec:matrix.org) room for clarification of any of these points.

- [x] Are [appropriate implementation(s)](https://spec.matrix.org/proposals/#implementing-a-proposal) specified in the MSC’s PR description?
- [x] Are all MSCs that this MSC depends on already accepted?
- [x] For each new endpoint that is introduced:
  - [x] Have authentication requirements been specified?
  - [x] Have rate-limiting requirements been specified?
  - [x] Have guest access requirements been specified?
  - [x] Are error responses specified?
    - [x] Does each error case have a specified `errcode` (e.g. `M_FORBIDDEN`) and HTTP status code?
      - [x] If a new `errcode` is introduced, is it clear that it is new?
- [x] Will the MSC require a new room version, and if so, has that been made clear?
  - [x] Is the reason for a new room version clearly stated? For example, modifying the set of redacted fields changes how event IDs are calculated, thus requiring a new room version.
- [x] Are backwards-compatibility concerns appropriately addressed?
- [x] Are the [endpoint conventions](https://spec.matrix.org/latest/appendices/#conventions-for-matrix-apis) honoured?
  - [x] Do HTTP endpoints `use_underscores_like_this`?
  - [x] Will the endpoint return unbounded data? If so, has pagination been considered?
  - [x] If the endpoint utilises pagination, is it consistent with [the appendices](https://spec.matrix.org/latest/appendices/#pagination)?
- [x] An introduction exists and clearly outlines the problem being solved. Ideally, the first paragraph should be understandable by a non-technical audience.
- [x] All outstanding threads are resolved
  - [x] All feedback is incorporated into the proposal text itself, either as a fix or noted as an alternative
- [x] While the exact sections do not need to be present, the details implied by the proposal template are covered. Namely:
  - [x] Introduction
  - [x] Proposal text
  - [x] Potential issues
  - [x] Alternatives
  - [x] Dependencies
- [x] Stable identifiers are used throughout the proposal, except for the unstable prefix section
  - [x] Unstable prefixes [consider](https://github.com/matrix-org/matrix-spec-proposals/blob/main/README.md#unstable-prefixes) the awkward accepted-but-not-merged state
  - [x] Chosen unstable prefixes do not pollute any global namespace (use “org.matrix.mscXXXX”, not “org.matrix”).
- [x] Changes have applicable [Sign Off](https://github.com/matrix-org/matrix-spec-proposals/blob/main/CONTRIBUTING.md#sign-off) from all authors/editors/contributors
- [x] There is a dedicated "Security Considerations" section which detail any possible attacks/vulnerabilities this proposal may introduce, even if this is "None.". See [RFC3552](https://datatracker.ietf.org/doc/html/rfc3552) for things to think about, but in particular pay attention to the [OWASP Top Ten](https://owasp.org/www-project-top-ten/).

clokep commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,68 @@
+Consumers of policy lists SHOULD note that recommendations can still appear 
+outside of `m.policy` rooms. Existing rooms probably won't have the new room 
+type, and some communities might mix conversation and recommendations in the 
+same room (therefore not dedicating the room to recommendations).
```
The current text is confusing. I would make it clear that `m.policy` rooms are not expected to have conversation.

Maybe use a paragraph break to talk about legacy policy rooms and that nothing in this MSC applies to them?

turt2live replied:
Would be resolved by https://github.com/matrix-org/matrix-spec-proposals/pull/3784#discussion_r3050235722

turt2live replied:
I'm assuming the concern was also supposed to be resolved.

@mscbot resolve Confusing behavior around whether or not conversation occurs in policy rooms.

clokep replied:
Yes, thank you!

turt2live commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,68 @@
+This change is largely informative and carries no direct security impact. Clients 
+which interpret the `m.policy` room type will need to consider security in their
+implementations. For example, if hiding the rooms then notifications in those 
+rooms will need some consideration.
```
Spaces left whether a conversation can happen in the room as an implementation detail because it was theorized as possibly useful.

turt2live replied:
I believe this may be mostly/all resolved by https://github.com/matrix-org/matrix-spec-proposals/pull/3784#discussion_r3050235722

turt2live commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,68 @@
+Consumers of policy lists SHOULD note that recommendations can still appear 
+outside of `m.policy` rooms. Existing rooms probably won't have the new room 
+type, and some communities might mix conversation and recommendations in the 
+same room (therefore not dedicating the room to recommendations).
```
I feel the suggestion implies the lack of type sufficiently. This will need line wrapping, but is otherwise great - thank you! 

@clokep @Johennes - would this resolve your concerns around the MSC not spelling it out?

clokep replied:
Yeah, I think that is clearer.

Johennes replied:
Yes, thanks.

turt2live replied:
@FSG-Cat this looks like it'd unblock the MSC.

FSG-Cat replied:
Suggestion with line wrapping has been applied.

turt2live commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,68 @@
+left as an implementation detail. 
+
+Consumers of policy lists SHOULD note that recommendations can still appear 
+outside of `m.policy` rooms. Existing rooms probably won't have the new room 
```
I believe this would be resolved by https://github.com/matrix-org/matrix-spec-proposals/pull/3784#discussion_r3050235722

FSG-Cat committed:
Fix markdown linting issues.

This is a separate commit from the proper review feedback commit so that the feedback integration isnt drowned in noise.

FSG-Cat committed:
Integrate review feedback from Gnuxie

Co-authored-by: Gnuxie <Gnuxie@users.noreply.github.com>

clokep commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -22,10 +22,9 @@ Rooms dedicated to containing policy list recommendations SHOULD use a newly
-outside of `m.policy` rooms. Existing rooms probably won't have the new room
-type, and some communities might mix conversation and recommendations in the
-same room (therefore not dedicating the room to recommendations).
+Where the `m.policy` room type is used, conversation is not expected.
```
Small quibble that I wouldn't block this MSC on, but should we use `m.moderation_policy` to be a bit more descriptive? I could think of multiple types of policies.

TheArcaneBrony replied:
Shouldn't the policy server MSC then also be updated to use `m.moderation_policy` as an event type rather than `m.policy` or whatever the final variant was?

FSG-Cat replied:
Im not married to any particular final name for the room type. I was just copying the precedent set by spaces at the time of writing originally.

So i will take this to the saftey guild and ask them for input on what the name should be.

clokep replied:
Maybe we're already using this term, and that's ok. Thanks for providing the context on why this was chosen. It just stuck out to me as really vague.

Would be curious for other SCT input here.

richvdh replied:
I feel like @turt2live would have more context on whether this is a sensible name or not. I agree that `m.policy` feels a bit vague.

turt2live replied:
As the person who advocated for `m.room.policy` to enter the spec as-is, I'm not sure my opinion is super reliable 😅. I'm personally fine with `m.policy` as a room type because it's "a room containing a policy". The 'policy' happens to be expressed with primarily ban rules, but in future could be more than that (anonymous reputation data, shades of gray policy definitions to determine on-topic-ness, etc). 

I'm also generally of the opinion that calling it a "moderation policy list" in the spec is wrong, but I'm not concerned enough to actually raise a bug fix PR or issue for it. The use case is to communicate general policy data, not specific to moderation. This could be a security policy, entrance policy, community volunteer policy, etc.

FSG-Cat replied:
A practical use case that exists today where policies are not "moderation policies" would actually be the Access controls for Mjolnir and Draupnir Appservice mode. While they are broken in Draupnir currently they work in Mjolnir as far as i know because it used to work in Draupnir and we did not touch that part of the codebase to my knowledge in that way. (I should note we use the moderation policy recommendations in this context but thats also more of a easiest path forward type of situation i would argue. But thats also only for the disallow policy the allow policy is a custom one in the org.matrix namespace.)

In this case the policies simply control if your bot is allowed to start or not and if your doing Appservice hosting as a service you can easily hook this into a subscription management system. Be that paid or free subscriptions.

Point im trying to make is that we have a concrete already in the wild use case where policies are not strict moderation but something a bit different and we have lots of exploration wanting to extend this system further.

Personally i would say that room sub typing also comes into play here as if we had sub typing then definitively could argue if moderation should be its own subtype and the flagship subtype for the first MSC to try to define a policy subtype due to its clear and proven track record.

Now the fact that the appservice admin room is where Mjolnir and Draupnir stores the policy list is a design choice that as far as i understand exists out of convenience not any actual necessity. Appservice mode was a small beta project that had beta issues that make complete sense in context. The admin room having the policies is one of them and is a choice that i personally would argue for moving away from if one wants to modernize the mode (and is something i may personally look at in the future inside of the Draupnir project) as the room type benefits are quite large for specialized tools like RMU and any other matrix client that wants to implement policy manipulation.

richvdh commented on proposals/3784-using-room-type-of-m.policy-for-policy-rooms.md:
```diff
@@ -0,0 +1,65 @@
+After this MSC gets merged if a stakeholder has elected to remove its support for the unstable prefix if any
+Because the room type is immutable, rooms which use the unstable room type might
+find themselves "unsupported" when/if implementations drop support for the unstable
+type when this proposal becomes stable.
```
I am utterly unable to parse this sentence.

FSG-Cat replied:
This sentence is meant to read in a simplified form as. If a client supports the unstable room type and later removes it then rooms that used to be "supported" as in recognized are no longer recognized.

The practical example is that if you try to look up policy lists using a client that does not support the unstable version you wont find a single policy list in all of matrix via room type today as no policy lists with the stable room type exist.

richvdh replied:
Can you please fix the MSC so that it can be understood?

Note that my comment applies to all four lines "After this ... becomes stable". It seems to be one huge sentence.

FSG-Cat replied:
A updated version of the text in question has been pushed that hopefully resolves this concern.

richvdh replied:
thanks. much clearer now.

FSG-Cat committed:
Reword Unstable section to be more readable.

turt2live commented:
Spec PR: https://github.com/matrix-org/matrix-spec/pull/2411
